cdap-common/src/main/java/co/cask/cdap/common/zookeeper/ReentrantDistributedLock.java

```diff
@@ -0,0 +1,395 @@
+        // If the current id is the smallest one, nodeToWatch will be null
+        String nodeToWatch = findNodeToWatch(children, lockNode, guid);
+
+        // Step 3a. lockNode is the lowest, hence this become lock owner.
+        if (nodeToWatch == null) {
+          // Acquired lock
+          completion.set(lockNode);
+        } else if (!waitForLock) {
+          // This is for the case of tryLock() without timeout.
+          completion.cancel(true);
+        }
+        // If the lock acquisition is completed, due to whatever reason, we need to watch for any other nodes
+        if (completion.isDone()) {
+          return;
+        }
+
+        // Step 3b and 4. Call exists on the next lowest sequence ID
+        OperationFuture<Stat> existsFuture = zkClient.exists(nodeToWatch, new Watcher() {
+          @Override
+          public void process(WatchedEvent event) {
+            if (!completion.isDone()) {
+              // If the watching node changed, go to step 2.
+              doAcquire(completion, waitForLock, guid, lockNode);
+            }
+          }
+        });
+
+        // Step 5. Depends on the exists call result, either go to step 2 if the nodeToWatch is gone or just
+        // let the watcher to trigger step 2 when there is change to the nodeToWatch.
+        Futures.addCallback(existsFuture, new FutureCallback<Stat>() {
+          @Override
+          public void onSuccess(@Nullable Stat result) {
+            // Result == null if the node is no longer exists
+            if (result == null && !completion.isDone()) {
+              doAcquire(completion, waitForLock, guid, lockNode);
+            }
+          }
+
+          @Override
+          public void onFailure(Throwable t) {
+            // If failed to call exists, fail the lock acquisition
+            completion.setException(t);
+          }
+        });
+      }
+
+      @Override
+      public void onFailure(Throwable t) {
+        if (lockPath != null) {
+          completion.setException(t);
+        } else {
+          doAcquire(completion, waitForLock, guid, null);
+        }
+      }
+    });
+  }
+
+  /**
+   * Deletes the given node if the given future failed.
+   */
+  private void deleteNodeOnFailure(final ListenableFuture<?> future, final String node) {
+    future.addListener(new Runnable() {
+      @Override
+      public void run() {
+        try {
+          future.get();
+        } catch (Exception e) {
+          zkClient.delete(node);
+        }
+      }
+    }, Threads.SAME_THREAD_EXECUTOR);
+  }
+
+  /**
+   * Find the maximum node that is smaller than node of the given lockPath.
+   *
+   * @return the node found or {@code null} if no such node exist
+   */
+  private String findNodeToWatch(NodeChildren children, String lockPath, String guid) {
+    // Lock path is "path/guid-id"
+    int guidLen = guid.length();
+    int id = Integer.parseInt(lockPath.substring(path.length() + guidLen + 2));
```

Comment:
`+2` is for the `/` and `-`. Current logic has no assumption on the sequence id length, but probably can just use `length - 10` to make the logic simpler since it's in the ZK spec (unless they change it in some future version).